onnxruntime/contrib_ops/cpu/attnlstm/deep_cpu_attn_lstm.cc

```diff
@@ -38,15 +39,21 @@ DeepCpuAttnLstmOp::Compute(OpKernelContext* context) const {
   Status status;
   // auto& logger = context->Logger();
 
-  auto data_type = X.DataType();
-  if (data_type == DataTypeImpl::GetType<float>())
-    status = ComputeImpl<float>(*context);
-  else if (data_type == DataTypeImpl::GetType<double>()) {
-    /* Need to update all the helpers to support double...
+  auto prim_type = X.DataType()->AsPrimitiveDataType();
+  ORT_ENFORCE(prim_type  != nullptr, "Invalid data type for LSTM operator of ", X.DataType());
```

Comment:
This is pretty cryptic. Is this checking that X is a tensor, and we can get the element type for the tensor? If so, 'AsPrimitiveDataType' doesn't really convey that.

Reply:
Also the model loading should have validated that the types for all node inputs/outputs are valid, so possibly this check isn't needed.

---
In reply to: [341909252](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r341909252) [](ancestors = 341909252)

Reply:
I have a similar question: at first glance (of earlier code preceding this), I thought "PrimitiveDataType" was used to indicate the type of elements stored inside tensors (for example), while "TensorType" would indicate the type of a tensor. But this code fragment is slightly confusing. Perhaps it is the method name "DataType" as used in "X.DataType()" above that is confusing … I guess that is supposed to return the Datatype of an element of the tensor? May be if that was called "ElementDataType" it would be helpful.